llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -5701,13 +5702,19 @@ TargetLowering::ParseConstraints(const DataLayout &DL,
   return ConstraintOperands;
 }
 
-/// Return an integer indicating how general CT is.
-static unsigned getConstraintGenerality(TargetLowering::ConstraintType CT) {
+/// Return a number indicating our preference for chosing a type of constraint
+/// over another, for the purpose of sorting them. Immediates are almost always
+/// preferrable (when they can be emitted).
+/// FIXME: We should prefer registers over memory but doing so may lead to
+/// unrecoverable register exhaustion later.
+/// https://github.com/llvm/llvm-project/issues/20571
+static unsigned getConstraintPiority(TargetLowering::ConstraintType CT) {
   switch (CT) {
-  case TargetLowering::C_Immediate:
-  case TargetLowering::C_Other:
   case TargetLowering::C_Unknown:
     return 0;
+  case TargetLowering::C_Immediate:
+  case TargetLowering::C_Other:
+    return 4;
```

Comment:
I don't quite get what the numbers represent here. Is "higher" more preferable or is "lower"? If lower is preferable, then shouldn't `C_Immediate` be more like `C_Register` rather than `C_Other`? Also, please reorder the cases into the sort order...

Reply:
Higher is preferable; I'll add that info the method comment.  Immediate is preferrable to register. Other is an oddball, we use that for the flag constraints.

Reply:
Also, this change intentionally //increases// the priority for these 2; previously `chooseConstraint` would stop as soon as it hit one that could be lowered.  This is //conceptually// the same thing as having a higher priority.

Reply:
https://github.com/llvm/llvm-project/pull/67057/commits/a542e81ffc8a7d178f4a661c40c6155e04388e63